app/models/workflow.rb

```diff
@@ -18,4 +21,24 @@ def external_processing?
   def external_signal?
     template&.signal_setting.present?
   end
+
+  private
+
+  def table
+    self.class.arel_table
+  end
+
+  def new_sequence
+    return move_rest_higher if sequence
+
+    self.sequence = self.class.last&.sequence.to_i + 1
+  end
+
+  # move all related sequence number one higher if the desired number is in use
+  def move_rest_higher
+    return unless sequence && self.class.where(:sequence => sequence).exists
+
+    self.class.where(table[:sequence].gteq(sequence)).update_all("sequence = (-sequence - 1)")
+    self.class.where(table[:sequence].lt(0)).update_all("sequence = (-sequence)")
```

Comment:
should these two steps are `locked` to prevent concurrent issue?

Reply:
try to avoid locking by using transaction and retry.

Reply:
I saw it in workflow `update` only, but no in `create`, any reason?

Reply:
I saw it in workflow `update` only, but no in `create`, any reason?

Reply:
create by default will always place the new workflow to the end of the sequencing, unless the number is given by caller and this method does get called.

BTW, transaction is not needed in our service classes. See the document https://guides.rubyonrails.org/active_record_callbacks.html#halting-execution. I am going to remove them.

Reply:
👍